// @solana/transaction-introspection/dist/index.node.cjs
'use strict';

var codecsStrings = require('@solana/codecs-strings');
var errors = require('@solana/errors');
var transactionMessages = require('@solana/transaction-messages');
var transactions = require('@solana/transactions');
var instructions = require('@solana/instructions');

// src/decode-rpc-transaction.ts
var EMPTY_LOADED_ADDRESSES = { readonly: [], writable: [] };
function getLoadedAddresses(meta) {
  const loaded = meta?.loadedAddresses;
  return loaded ?? EMPTY_LOADED_ADDRESSES;
}
function decodeFromWire(wireBytes) {
  const transaction = transactions.getTransactionDecoder().decode(wireBytes);
  const compiledMessage = transactionMessages.getCompiledTransactionMessageDecoder().decode(transaction.messageBytes);
  return { compiledMessage, transaction };
}
function decodeFromBase64(rpcTx) {
  const [b64] = rpcTx.transaction;
  const { compiledMessage, transaction } = decodeFromWire(codecsStrings.getBase64Encoder().encode(b64));
  return { compiledMessage, loadedAddresses: getLoadedAddresses(rpcTx.meta), transaction };
}
function decodeFromBase58(rpcTx) {
  const [b58] = rpcTx.transaction;
  const { compiledMessage, transaction } = decodeFromWire(codecsStrings.getBase58Encoder().encode(b58));
  return { compiledMessage, loadedAddresses: getLoadedAddresses(rpcTx.meta), transaction };
}
function decodeFromJson(rpcTx) {
  const base58 = codecsStrings.getBase58Encoder();
  const { message } = rpcTx.transaction;
  const staticAccounts = [...message.accountKeys];
  const getCompiledInstructions = () => message.instructions.map((ix) => ({
    ...ix.accounts.length ? { accountIndices: [...ix.accounts] } : null,
    ...ix.data.length ? { data: base58.encode(ix.data) } : null,
    programAddressIndex: ix.programIdIndex
  }));
  const header = {
    numReadonlyNonSignerAccounts: message.header.numReadonlyUnsignedAccounts,
    numReadonlySignerAccounts: message.header.numReadonlySignedAccounts,
    numSignerAccounts: message.header.numRequiredSignatures
  };
  const version = "version" in rpcTx ? rpcTx.version : "legacy";
  let compiledMessage;
  switch (version) {
    case "legacy":
      compiledMessage = {
        header,
        instructions: getCompiledInstructions(),
        lifetimeToken: message.recentBlockhash,
        staticAccounts,
        version: "legacy"
      };
      break;
    case 0: {
      const addressTableLookups = "addressTableLookups" in message ? message.addressTableLookups.map((l) => ({
        lookupTableAddress: l.accountKey,
        readonlyIndexes: l.readonlyIndexes,
        writableIndexes: l.writableIndexes
      })) : [];
      compiledMessage = {
        ...addressTableLookups.length ? { addressTableLookups } : null,
        header,
        instructions: getCompiledInstructions(),
        lifetimeToken: message.recentBlockhash,
        staticAccounts,
        version: 0
      };
      break;
    }
    case 1: {
      const instructionData = message.instructions.map((ix) => base58.encode(ix.data));
      compiledMessage = {
        // The `'json'` encoding does not carry the v1 transaction
        // config, so the synthesized message always reports an
        // empty one.
        configMask: 0,
        configValues: [],
        header,
        instructionHeaders: message.instructions.map((ix, i) => ({
          numInstructionAccounts: ix.accounts.length,
          numInstructionDataBytes: instructionData[i].byteLength,
          programAccountIndex: ix.programIdIndex
        })),
        instructionPayloads: message.instructions.map((ix, i) => ({
          instructionAccountIndices: [...ix.accounts],
          instructionData: instructionData[i]
        })),
        lifetimeToken: message.recentBlockhash,
        numInstructions: message.instructions.length,
        numStaticAccounts: staticAccounts.length,
        staticAccounts,
        version: 1
      };
      break;
    }
    default: {
      const _exhaustiveCheck = version;
      throw new errors.SolanaError(errors.SOLANA_ERROR__TRANSACTION__VERSION_NUMBER_NOT_SUPPORTED, {
        unsupportedVersion: _exhaustiveCheck
      });
    }
  }
  return { compiledMessage, loadedAddresses: getLoadedAddresses(rpcTx.meta) };
}
function isBase64Response(rpcTx) {
  const t = rpcTx.transaction;
  return Array.isArray(t) && t[1] === "base64";
}
function isBase58Response(rpcTx) {
  const t = rpcTx.transaction;
  return Array.isArray(t) && t[1] === "base58";
}
function getJsonShapedMessage(rpcTx) {
  const t = rpcTx.transaction;
  if (typeof t !== "object" || t === null || Array.isArray(t) || !("message" in t)) return void 0;
  const message = t.message;
  if (!message || typeof message !== "object" || !Array.isArray(message.instructions)) return void 0;
  return message;
}
function isJsonResponse(rpcTx) {
  const message = getJsonShapedMessage(rpcTx);
  return message != null && typeof message.header === "object" && message.header !== null;
}
function isJsonParsedResponse(rpcTx) {
  const message = getJsonShapedMessage(rpcTx);
  return message != null && !("header" in message);
}
function decodeTransactionFromRpcResponse(rpcTx) {
  const tx = rpcTx;
  if (isBase64Response(tx)) return decodeFromBase64(tx);
  if (isBase58Response(tx)) return decodeFromBase58(tx);
  if (isJsonResponse(tx)) return decodeFromJson(tx);
  if (isJsonParsedResponse(tx)) {
    throw new errors.SolanaError(errors.SOLANA_ERROR__TRANSACTION_INTROSPECTION__CANNOT_DECODE_JSON_PARSED_TRANSACTION);
  }
  throw new errors.SolanaError(errors.SOLANA_ERROR__TRANSACTION_INTROSPECTION__UNRECOGNIZED_GET_TRANSACTION_RESPONSE);
}
function getInnerInstructionsFromMeta(meta, accountMetas) {
  if (!meta.innerInstructions) return [];
  const base58 = codecsStrings.getBase58Encoder();
  const result = [];
  for (const group of meta.innerInstructions) {
    for (let innerIndex = 0; innerIndex < group.instructions.length; innerIndex++) {
      const ix = group.instructions[innerIndex];
      const programMeta = accountMetas[ix.programIdIndex];
      if (!programMeta) {
        throw new errors.SolanaError(
          errors.SOLANA_ERROR__TRANSACTION__FAILED_TO_DECOMPILE_INSTRUCTION_PROGRAM_ADDRESS_NOT_FOUND,
          { index: ix.programIdIndex }
        );
      }
      const accounts = ix.accounts.map((i) => {
        const accountMeta = accountMetas[i];
        if (!accountMeta) {
          throw new errors.SolanaError(
            errors.SOLANA_ERROR__TRANSACTION__FAILED_TO_DECOMPILE_INSTRUCTION_ACCOUNT_INDEX_OUT_OF_RANGE,
            { index: i }
          );
        }
        return accountMeta;
      });
      const data = base58.encode(ix.data);
      result.push({
        ...accounts.length ? { accounts } : null,
        ...data.byteLength ? { data } : null,
        programAddress: programMeta.address,
        trace: {
          innerIndex,
          kind: "inner",
          outerIndex: group.index,
          ...ix.stackHeight != null ? { stackHeight: ix.stackHeight } : {}
        }
      });
    }
  }
  return result;
}
function getAccountMetasFromCompiledTransactionMessage(compiledMessage, loadedAddresses) {
  const { header, staticAccounts } = compiledMessage;
  const numWritableSignerAccounts = header.numSignerAccounts - header.numReadonlySignerAccounts;
  const numWritableNonSignerAccounts = staticAccounts.length - header.numSignerAccounts - header.numReadonlyNonSignerAccounts;
  const metas = [];
  let i = 0;
  for (let n = 0; n < numWritableSignerAccounts; n++, i++) {
    metas.push({ address: staticAccounts[i], role: instructions.AccountRole.WRITABLE_SIGNER });
  }
  for (let n = 0; n < header.numReadonlySignerAccounts; n++, i++) {
    metas.push({ address: staticAccounts[i], role: instructions.AccountRole.READONLY_SIGNER });
  }
  for (let n = 0; n < numWritableNonSignerAccounts; n++, i++) {
    metas.push({ address: staticAccounts[i], role: instructions.AccountRole.WRITABLE });
  }
  for (let n = 0; n < header.numReadonlyNonSignerAccounts; n++, i++) {
    metas.push({ address: staticAccounts[i], role: instructions.AccountRole.READONLY });
  }
  if (loadedAddresses) {
    for (const address of loadedAddresses.writable) {
      metas.push({ address, role: instructions.AccountRole.WRITABLE });
    }
    for (const address of loadedAddresses.readonly) {
      metas.push({ address, role: instructions.AccountRole.READONLY });
    }
  }
  return metas;
}
function getInstructionsFromCompiledTransactionMessage(compiledMessage, loadedAddresses) {
  const metas = getAccountMetasFromCompiledTransactionMessage(compiledMessage, loadedAddresses);
  return normalizeCompiledInstructions(compiledMessage).map((ix) => resolveInstruction(ix, metas));
}
function getInstructionsFromCompiledTransactionMessageWithMetas(compiledMessage, accountMetas) {
  return normalizeCompiledInstructions(compiledMessage).map((ix) => resolveInstruction(ix, accountMetas));
}
function normalizeCompiledInstructions(compiledMessage) {
  if (compiledMessage.version === "legacy" || compiledMessage.version === 0) {
    return compiledMessage.instructions.map((ix) => ({
      accountIndices: ix.accountIndices ?? [],
      data: ix.data ?? new Uint8Array(),
      programAddressIndex: ix.programAddressIndex
    }));
  }
  if (compiledMessage.version === 1) {
    const { instructionHeaders, instructionPayloads } = compiledMessage;
    if (instructionHeaders.length !== instructionPayloads.length) {
      throw new errors.SolanaError(errors.SOLANA_ERROR__TRANSACTION__INSTRUCTION_HEADERS_PAYLOADS_MISMATCH, {
        numInstructionHeaders: instructionHeaders.length,
        numInstructionPayloads: instructionPayloads.length
      });
    }
    return instructionHeaders.map((header, i) => ({
      accountIndices: instructionPayloads[i].instructionAccountIndices,
      data: instructionPayloads[i].instructionData,
      programAddressIndex: header.programAccountIndex
    }));
  }
  const _exhaustiveCheck = compiledMessage;
  throw new errors.SolanaError(errors.SOLANA_ERROR__TRANSACTION__VERSION_NUMBER_NOT_SUPPORTED, {
    unsupportedVersion: _exhaustiveCheck.version
  });
}
function resolveInstruction(ix, metas) {
  const programMeta = metas[ix.programAddressIndex];
  if (!programMeta) {
    throw new errors.SolanaError(errors.SOLANA_ERROR__TRANSACTION__FAILED_TO_DECOMPILE_INSTRUCTION_PROGRAM_ADDRESS_NOT_FOUND, {
      index: ix.programAddressIndex
    });
  }
  const accounts = ix.accountIndices.map((i) => {
    const accountMeta = metas[i];
    if (!accountMeta) {
      throw new errors.SolanaError(
        errors.SOLANA_ERROR__TRANSACTION__FAILED_TO_DECOMPILE_INSTRUCTION_ACCOUNT_INDEX_OUT_OF_RANGE,
        {
          index: i
        }
      );
    }
    return accountMeta;
  });
  return {
    ...accounts.length ? { accounts } : null,
    ...ix.data.byteLength ? { data: ix.data } : null,
    programAddress: programMeta.address
  };
}

// src/walk-instructions.ts
function walkInstructions(args) {
  const { compiledMessage, loadedAddresses, meta } = args;
  const accountMetas = getAccountMetasFromCompiledTransactionMessage(compiledMessage, loadedAddresses);
  const outerInstructions = getInstructionsFromCompiledTransactionMessageWithMetas(compiledMessage, accountMetas);
  const innerByOuterIndex = /* @__PURE__ */ new Map();
  if (meta) {
    for (const inner of getInnerInstructionsFromMeta(meta, accountMetas)) {
      if (inner.trace.kind !== "inner") continue;
      const group = innerByOuterIndex.get(inner.trace.outerIndex);
      if (group) {
        group.push(inner);
      } else {
        innerByOuterIndex.set(inner.trace.outerIndex, [inner]);
      }
    }
  }
  const result = [];
  outerInstructions.forEach((instruction, index) => {
    result.push({ ...instruction, trace: { index, kind: "outer" } });
    const group = innerByOuterIndex.get(index);
    if (group) {
      result.push(...group);
      innerByOuterIndex.delete(index);
    }
  });
  for (const group of innerByOuterIndex.values()) {
    result.push(...group);
  }
  return result;
}

exports.decodeTransactionFromRpcResponse = decodeTransactionFromRpcResponse;
exports.getAccountMetasFromCompiledTransactionMessage = getAccountMetasFromCompiledTransactionMessage;
exports.getInnerInstructionsFromMeta = getInnerInstructionsFromMeta;
exports.getInstructionsFromCompiledTransactionMessage = getInstructionsFromCompiledTransactionMessage;
exports.walkInstructions = walkInstructions;
