// @solana/subscribable/dist/index.node.mjs
import { setMaxListeners } from 'events';
import { getAbortablePromise } from '@solana/promises';
import { SolanaError, SOLANA_ERROR__INVARIANT_VIOLATION__SUBSCRIPTION_ITERATOR_STATE_MISSING, SOLANA_ERROR__INVARIANT_VIOLATION__SUBSCRIPTION_ITERATOR_MUST_NOT_POLL_BEFORE_RESOLVING_EXISTING_MESSAGE_PROMISE } from '@solana/errors';

// ../event-target-impl/dist/index.node.mjs
var e = class extends globalThis.AbortController {
  constructor(...t) {
    super(...t), setMaxListeners(Number.MAX_SAFE_INTEGER, this.signal);
  }
};
var s = class extends globalThis.EventTarget {
  constructor(...t) {
    super(...t), setMaxListeners(Number.MAX_SAFE_INTEGER, this);
  }
};
var IDLE_STATE = Object.freeze({
  data: void 0,
  error: void 0,
  status: "idle"
});
function createReactiveActionStore(fn) {
  let state = IDLE_STATE;
  let currentController;
  const listeners = /* @__PURE__ */ new Set();
  function setState(next) {
    if (state.status === next.status && state.data === next.data && state.error === next.error) {
      return;
    }
    state = next;
    listeners.forEach((listener) => listener());
  }
  const dispatchAsyncWithSignal = async (userSignal, ...args) => {
    currentController?.abort();
    if (userSignal?.aborted) {
      setState({ data: state.data, error: userSignal.reason, status: "error" });
      throw userSignal.reason;
    }
    const controller = new e();
    currentController = controller;
    const signal = userSignal ? AbortSignal.any([controller.signal, userSignal]) : controller.signal;
    const previousData = state.data;
    const previousError = state.error;
    setState({ data: previousData, error: previousError, status: "running" });
    try {
      const result = await getAbortablePromise(fn(signal, ...args), signal);
      if (signal.aborted) {
        throw signal.reason;
      }
      setState({ data: result, error: void 0, status: "success" });
      return result;
    } catch (error) {
      if (controller.signal.aborted) {
        throw controller.signal.reason;
      }
      setState({ data: previousData, error, status: "error" });
      throw error;
    }
  };
  const dispatchAsync = (...args) => dispatchAsyncWithSignal(void 0, ...args);
  const dispatch = (...args) => {
    dispatchAsync(...args).catch(() => {
    });
  };
  return {
    dispatch,
    dispatchAsync,
    getState: () => state,
    reset: () => {
      currentController?.abort();
      currentController = void 0;
      setState(IDLE_STATE);
    },
    subscribe: (listener) => {
      listeners.add(listener);
      return () => {
        listeners.delete(listener);
      };
    },
    withSignal: (signal) => ({
      dispatch: (...args) => {
        dispatchAsyncWithSignal(signal, ...args).catch(() => {
        });
      },
      dispatchAsync: (...args) => dispatchAsyncWithSignal(signal, ...args)
    })
  };
}
var EXPLICIT_ABORT_TOKEN;
function createExplicitAbortToken() {
  return Symbol(
    process.env.NODE_ENV !== "production" ? "This symbol is thrown from a socket's iterator when the connection is explicitly aborted by the user" : void 0
  );
}
var UNINITIALIZED = Symbol();
function createAsyncIterableFromDataPublisher({
  abortSignal,
  dataChannelName,
  dataPublisher,
  errorChannelName
}) {
  const iteratorState = /* @__PURE__ */ new Map();
  function publishErrorToAllIterators(reason) {
    for (const [iteratorKey, state] of iteratorState.entries()) {
      if (state.__hasPolled) {
        iteratorState.delete(iteratorKey);
        state.onError(reason);
      } else {
        state.publishQueue.push({
          __type: 1 /* ERROR */,
          err: reason
        });
      }
    }
  }
  const abortController = new e();
  abortSignal.addEventListener("abort", () => {
    abortController.abort();
    publishErrorToAllIterators(EXPLICIT_ABORT_TOKEN ||= createExplicitAbortToken());
  });
  const options = { signal: abortController.signal };
  let firstError = UNINITIALIZED;
  dataPublisher.on(
    errorChannelName,
    (err) => {
      if (firstError === UNINITIALIZED) {
        firstError = err;
        abortController.abort();
        publishErrorToAllIterators(err);
      }
    },
    options
  );
  dataPublisher.on(
    dataChannelName,
    (data) => {
      iteratorState.forEach((state, iteratorKey) => {
        if (state.__hasPolled) {
          const { onData } = state;
          iteratorState.set(iteratorKey, { __hasPolled: false, publishQueue: [] });
          onData(data);
        } else {
          state.publishQueue.push({
            __type: 0 /* DATA */,
            data
          });
        }
      });
    },
    options
  );
  return {
    async *[Symbol.asyncIterator]() {
      if (abortSignal.aborted) {
        return;
      }
      if (firstError !== UNINITIALIZED) {
        throw firstError;
      }
      const iteratorKey = Symbol();
      iteratorState.set(iteratorKey, { __hasPolled: false, publishQueue: [] });
      try {
        while (true) {
          const state = iteratorState.get(iteratorKey);
          if (!state) {
            throw new SolanaError(SOLANA_ERROR__INVARIANT_VIOLATION__SUBSCRIPTION_ITERATOR_STATE_MISSING);
          }
          if (state.__hasPolled) {
            throw new SolanaError(
              SOLANA_ERROR__INVARIANT_VIOLATION__SUBSCRIPTION_ITERATOR_MUST_NOT_POLL_BEFORE_RESOLVING_EXISTING_MESSAGE_PROMISE
            );
          }
          const publishQueue = state.publishQueue;
          try {
            if (publishQueue.length) {
              state.publishQueue = [];
              for (const item of publishQueue) {
                if (item.__type === 0 /* DATA */) {
                  yield item.data;
                } else {
                  throw item.err;
                }
              }
            } else {
              yield await new Promise((resolve, reject) => {
                iteratorState.set(iteratorKey, {
                  __hasPolled: true,
                  onData: resolve,
                  onError: reject
                });
              });
            }
          } catch (e2) {
            if (e2 === (EXPLICIT_ABORT_TOKEN ||= createExplicitAbortToken())) {
              return;
            } else {
              throw e2;
            }
          }
        }
      } finally {
        iteratorState.delete(iteratorKey);
      }
    }
  };
}

// src/data-publisher.ts
function getDataPublisherFromEventEmitter(eventEmitter) {
  return {
    on(channelName, subscriber, options) {
      function innerListener(ev) {
        if (ev instanceof CustomEvent) {
          const data = ev.detail;
          subscriber(data);
        } else {
          subscriber();
        }
      }
      eventEmitter.addEventListener(channelName, innerListener, options);
      return () => {
        eventEmitter.removeEventListener(channelName, innerListener);
      };
    }
  };
}

// src/demultiplex.ts
function demultiplexDataPublisher(publisher, sourceChannelName, messageTransformer) {
  let innerPublisherState;
  const eventTarget = new s();
  const demultiplexedDataPublisher = getDataPublisherFromEventEmitter(eventTarget);
  return {
    ...demultiplexedDataPublisher,
    on(channelName, subscriber, options) {
      if (!innerPublisherState) {
        const innerPublisherUnsubscribe = publisher.on(sourceChannelName, (sourceMessage) => {
          const transformResult = messageTransformer(sourceMessage);
          if (!transformResult) {
            return;
          }
          const [destinationChannelName, message] = transformResult;
          eventTarget.dispatchEvent(
            new CustomEvent(destinationChannelName, {
              detail: message
            })
          );
        });
        innerPublisherState = {
          dispose: innerPublisherUnsubscribe,
          numSubscribers: 0
        };
      }
      innerPublisherState.numSubscribers++;
      const unsubscribe = demultiplexedDataPublisher.on(channelName, subscriber, options);
      let isActive = true;
      function handleUnsubscribe() {
        if (!isActive) {
          return;
        }
        isActive = false;
        options?.signal.removeEventListener("abort", handleUnsubscribe);
        innerPublisherState.numSubscribers--;
        if (innerPublisherState.numSubscribers === 0) {
          innerPublisherState.dispose();
          innerPublisherState = void 0;
        }
        unsubscribe();
      }
      options?.signal.addEventListener("abort", handleUnsubscribe);
      return handleUnsubscribe;
    }
  };
}

// src/reactive-stream-store.ts
var IDLE_STATE2 = Object.freeze({
  data: void 0,
  error: void 0,
  status: "idle"
});
function createReactiveStoreFromDataPublisherFactory({
  createDataPublisher,
  dataChannelName,
  errorChannelName
}) {
  let currentState = IDLE_STATE2;
  let currentInnerController;
  const subscribers = /* @__PURE__ */ new Set();
  function notify() {
    subscribers.forEach((cb) => cb());
  }
  function setState(next) {
    if (currentState.status === next.status && currentState.data === next.data && currentState.error === next.error) {
      return;
    }
    currentState = next;
    notify();
  }
  function performConnect(callerSignal) {
    currentInnerController?.abort();
    if (callerSignal?.aborted) {
      setState({ data: currentState.data, error: callerSignal.reason, status: "error" });
      return;
    }
    setState({ data: currentState.data, error: currentState.error, status: "loading" });
    const innerController = new e();
    currentInnerController = innerController;
    const signal = callerSignal ? AbortSignal.any([innerController.signal, callerSignal]) : innerController.signal;
    if (callerSignal) {
      callerSignal.addEventListener(
        "abort",
        () => {
          if (innerController.signal.aborted) return;
          setState({ data: currentState.data, error: callerSignal.reason, status: "error" });
          innerController.abort(callerSignal.reason);
        },
        { signal: innerController.signal }
      );
    }
    createDataPublisher(signal).then(
      (publisher) => {
        if (signal.aborted) return;
        publisher.on(
          dataChannelName,
          (data) => {
            setState({ data, error: void 0, status: "loaded" });
          },
          { signal }
        );
        publisher.on(
          errorChannelName,
          (err) => {
            if (currentState.status === "error") return;
            setState({ data: currentState.data, error: err, status: "error" });
            innerController.abort(err);
          },
          { signal }
        );
      },
      (err) => {
        if (signal.aborted) return;
        setState({ data: currentState.data, error: err, status: "error" });
        innerController.abort(err);
      }
    );
  }
  function performReset() {
    currentInnerController?.abort();
    currentInnerController = void 0;
    setState(IDLE_STATE2);
  }
  return {
    connect() {
      performConnect(void 0);
    },
    getState() {
      return currentState;
    },
    reset: performReset,
    subscribe(callback) {
      subscribers.add(callback);
      return () => {
        subscribers.delete(callback);
      };
    },
    withSignal(signal) {
      return {
        connect() {
          performConnect(signal);
        }
      };
    }
  };
}

export { createAsyncIterableFromDataPublisher, createReactiveActionStore, createReactiveStoreFromDataPublisherFactory, demultiplexDataPublisher, getDataPublisherFromEventEmitter };
